lib/events/stream.go

```diff
@@ -0,0 +1,694 @@
+	// If no context is passed in (like in tests) then set a background context.
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	return &StreamManager{
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: "stream",
+		}),
+		pool: sync.Pool{
+			New: func() interface{} {
+				return new(bytes.Buffer)
+			},
+		},
+		semaphore:    make(chan struct{}, defaults.ConcurrentStreams),
+		closeContext: ctx,
+	}
+}
+
+// ProcessStream reads events off a GRPC stream and processes them.
+func (s *StreamManager) ProcessStream(serverID string, uploader SessionUploader, stream proto.AuthService_StreamSessionRecordingServer) (*Stream, error) {
+	// Wrap the parent process shutdown context and create an upload context that is
+	// used by the uploader to cancel a upload if an error occurs while creating
+	// the archive.
+	ctx, cancel := context.WithCancel(s.closeContext)
+
+	st := &Stream{
+		manager:       s,
+		stream:        stream,
+		chunkType:     typeInit,
+		uploader:      uploader,
+		serverID:      serverID,
+		uploadContext: ctx,
+		uploadCancel:  cancel,
+		waitCh:        make(chan error),
+	}
+
+	// Start reading off the stream and processing events.
+	go st.start()
+
+	return st, nil
+}
+
+// takeSemaphore will acquire the semaphore.
+func (s *StreamManager) takeSemaphore() error {
+	select {
+	case s.semaphore <- struct{}{}:
+		return nil
+	case <-s.closeContext.Done():
+		return errContext
+	}
+}
+
+// releaseSemaphore will release the semaphore.
+func (s *StreamManager) releaseSemaphore() error {
+	select {
+	case <-s.semaphore:
+		return nil
+	case <-s.closeContext.Done():
+		return errContext
+	}
+}
+
+// Wait will block until the stream has been processed and a response written
+// to the client.
+func (s *Stream) Wait() error {
+	return <-s.waitCh
+}
+
+// Close will terminate the stream and write a response to the client.
+func (s *Stream) Close(err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Either the upload is complete (call cancel to free resources) or an error
+	// has occured (terminate upload), either way cancel the context.
+	s.uploadCancel()
+
+	// Close any resources (if not already closed) that may have been allocated.
+	if s.reader != nil {
+		s.reader.Close()
+	}
+	if s.writer != nil {
+		s.writer.Close()
+	}
+	if s.tarWriter != nil {
+		s.tarWriter.Close()
+	}
+	if s.zipWriter != nil {
+		s.zipWriter.Close()
+	}
+
+	// Send a response to the client with the result of the stream.
+	if err != nil {
+		err = s.stream.SendAndClose(&proto.StreamSessionResponse{
+			Success: false,
+			Message: err.Error(),
+		})
+	} else {
+		err = s.stream.SendAndClose(&proto.StreamSessionResponse{
+			Success: true,
+		})
+	}
+	if err != nil {
+		s.manager.log.Debugf("Failed to close stream %v: %v.", s.sessionID, err)
+	}
+
+	// Unblock the waiter with the result.
+	s.waitCh <- trace.Wrap(err)
+}
+
+// start pulls events off the stream and processes them.
+func (s *Stream) start() {
+	for {
```

Comment:
I would add `defer cancel` here, because it seems like in case if err == io.EOF, Wait will never return

Reply:
That was intentional, Wait blocks until the upload is complete so the client can be informed if the upload was successful or not.